### PR TITLE
Extensions infrastructure

### PR DESCRIPTION
**Pull request by uliska**
I have finally started to work on an extension infrastructure allowing to integrate external code in Frescobaldi (which has been discussed in #599, #592, #1069 and probably elsewhere too). This start was motivated by personal need to be able to add custom (partially private) functionality without having to actually fork the repository.

This PR is as small as it could be. I intended to propose the changes as early as possible, but it turned out that it was necessary to create an essentially working version for a first step to make any sense. But I hope it goes in the right direction.

In order to test this PR you have to do the following (in addition to checking out the branch, of course):

* Create one arbitrary directory that will hold all future extensions (e.g. `~/git/frescobaldi/extensions`)
* Clone https://github.com/frescobaldi-extensions/sample-extension.git into a subdirectory of that folder
* Go to the new "Extensions" page in Frescobaldi's Preferences dialog and store the above extensions directory.

Then there should be an `Tools=>Extensions` submenu available where all extensions will expose their (optional) Tool panels and additional actions:

grafik

Custom selections from the available actions are made available in the context menus of the editor and the music view (so far, other hooks to expose extension functionality in the GUI should be discussed and implemented later).

Both the files in the `extensions` module and the sample extension are heavily commented to help understand what I did.

Basically an extension is a Python package that implements an `Extension` class in its `__init__.py`. Optionally it can provide an action collection and a tool panel (from which at least one must be present). For all required classes appropriate base classes are available that bridge between the usual Frescobaldi classes and the extension, in addition a number of properties are passed through for the extension to directly access (e.g. the current document or its cursor).
The basic tasks of creating an extension are very small and simple, and then the extension has full access to Frescobaldi's modules and functionality (I assume we should place an appropriate warning somewhere to only load extensions from reliable sources because they will have unrestricted (non-root) access to the computer, but I don't think we should actually try to implement security like a sandbox). 

I hope this approach goes into the right direction and we can rather quickly iron out any issues to be able to merge this as a first iteration of an extension infrastructure. So far I have the impression it is already very powerful and runs smoothly.

---

Some issues I'm already aware of or that I haven't tackled yet:

The first group should probably be fixed before merging this Pull Request:

* [x] I want the Preference page to display information about the extensions (that has to be provided by some config/metadata file), at least the loaded extensions should be listed by name. Preliminary implementation in 195641407a54a2908f0e049a739082265fabda03 (should be sufficient for now)
* [x] extensions should be able to load icons from their directory. An extension icon (like a favicon) should be used, in the Preferences page and maybe at some point in the extension Panel.  
Implemented in 7c88b1f249cae6ab3df873c2791c5fad2af6780f and a6402b4dbe8bdb1378f8d4fcdcdae154a8e02431
* [x] There are glitches in the handling of settingsChanged when the root path is changed from a valid to an invalid path and back. In this case the menus have to be reloaded, and it seems that isn't done perfectly yet.  
The same is true (actually worse) when de/activating extensions altogether. Therefore this functionality is currently disabled.  
I've decided to require a restart in these cases. We don't know how (3rd-party) extensions will behave, and it is impossible to "unload" a module from the running application. edf93d82bc18d3bbac8f895ad9308a3144d3ad49
* [ ] I'm not sure if I'm correctly handling the reference to the main window. There are provisions in the code to manage more than one mainwindow instance, and I have never really understood them. IIUC the Extensions object (`app.extensions()`) should be a global one, so I'm not sure what would happen if someone actually has multiple MainWindow instances.

The others may be deferred to later development (and should be added as issues):

* [ ] It would be good to extend the localization support to extensions.
* [ ] It should be possible to assign keyboard shortcuts to extensions' actions. This must gracefully handle the case when extensions are *not* available anymore.
* [x] The preferences should allow the deactivation of extensions in general 
* [x] It should be possible to deactivate individual extensions (c840b4f24c9962c6f3a5012174ec9c2296f7d4fd)
* [ ] Also there should be a means for custom sorting of extensions (how they appear in the menus)
* [x] Probably a mechanism is needed to ensure version compatibility, say, an API version number extensions can reliably refer to.
* [ ] Integrate documentation. E.g. force an extension to provide a What's This item. Or automatically integrate extension documentation in the User guide

### CODE REVIEW

**Comment by uliska** [comment]
Hey, as you can see from the commit list I have continued working on this branch.
I've made the code more straightforward (I hope), and especially made it easier to create extensions. Also I've implemented a number of the issues I listed when originally opening the Pull Request, specifically extensions can now provide a configuration widget that is implicitly shown in the Preferences.

In parallel I've converted a custom Tool panel that I use for a private project to an extension, and it turns out that the system seems to work pretty smoothly already.

In addition to the heavily commented sample extension I intend to create another one called "tutorial". This will be much less commented (the comments actually hide the fact how simple an extension can be) but accompanied by a proper written introduction.

What I'm not completely sure about is how extensively the extension mechanism should be reflrected in the user guide.

But in general I'd say this could be ready to be merged now.

*Preferences showing the configuration widget of the sample extension:*

grafik

*Action reflecting the option set in the Preferences:*

grafik

**Comment by uliska** [comment]
I've now added handling of extensions's meta information, proper (I hope) handling of faulty extensions, and the userguide pages. The user guide is *exclusively* from the perspective of the user, a guide how to *write* extensions should not be added to the userguide at all. I'll write a separate document that can be either added to the (new) website or provided as a PDF resource in the `sample-extension` repository.

I think this is now ready for a final review, any further wishes could be added later IMO. The aspect I'm not 100% comfortable with is (as noted in the bullet list in the original description) the handling of the `mainwindow`.

**Comment by uliska** [comment]
To test the error handling you can do the following:

* Add a directory with arbitrary files to the extension directory  
`=>` should be ignored
* Add an empty `extension.cnf` to that directory  
`=>` A warning is displayed, information is present in the Preference page  
`=>` Additionally errors about not loaded extension
* Change a mandatory field in the `extension.cnf` of an existing extension (comment out `extension-name`)  
`=>` similar behaviour, but extension loads
* Cause an error in an extension (e.g. `import whatever` in its `__init__py`)  
`=>`extension doesn't load

**Comment by uliska** [comment]
Further improvement: significantly simplify handling of settings (and preventing extensions from messing around with the gobal settings/preferences).

**Comment by uliska** [comment]
Minor polishing and small additions while re-creating a fairly complex private extension. I really think this is good to go now.

**Comment by uliska** [comment]
After quite some pause I returned to this branch and made some final (?) improvements and polishing.

I will now also complete the mentioned custom extension and see if this indicates the need for more changes, but since there seems noone ready to test this I'll take the liberty to eventually merge this branch without explicit review.

**Comment by PeterBjuhr** [comment]
I'll try to manage to review it this weekend! But it's pretty comprehensive... :slightly_smiling_face: 


.

**Comment by uliska** [comment]
Indeed. I'm sorry about that but it wouldn't have been possible to do it differently. Originally I had intended to *only* prepare some very first steps, but it turned out to become a pretty comprehensive solution that I have just bumped to a `0.9.0` API version statement ;-)

It would have probably been better if this had been developed by a team/pair, but I somehow feel pretty confident that I didn't produce too much nonsense ...

File: frescobaldi_app/app.py [hunk 1]
```diff
@@ -228,6 +228,21 @@ def displayhook(obj):
     if obj is not None:
         print(repr(obj))
 
+
+_extensions = None
+
+def load_extensions(mainwindow):
+    """Called from MainWindow.__init__()"""
+    import extensions
+    global _extensions
```
Comment (PeterBjuhr) [inline frescobaldi_app/app.py]: The use of global variables gives me a warning sign. But in this case it's following the rest of the code and could perhaps be seen as a constant object?

File: frescobaldi_app/extensions/__init__.py [hunk 2]
```diff
@@ -0,0 +1,914 @@
...
+        """Returns a reference to the main window."""
+        return self.parent().mainwindow()
+
+    def text_cursor(self):
+        """The text cursor of the current document."""
+        return self.mainwindow().textCursor()
+
+
+class ExtensionMixin(QObject):
+    """Convenience mixin class for widgets in an extension.
```
Comment (PeterBjuhr) [inline frescobaldi_app/extensions/__init__.py]: @uliska Sorry I don't have the time to dig into the details of the code! Could you reply more about why you are choosing to use both mixins and inheritance?

File: frescobaldi_app/app.py (hunk 1 above)
Comment (uliska) [reply frescobaldi_app/app.py]: Yes, that was the rationale: implement the caching like with other comparable objects.

File: frescobaldi_app/extensions/__init__.py (hunk 2 above)
Comment (uliska) [reply frescobaldi_app/extensions/__init__.py]: I came to the conclusion that the functionality provided by this class had to be available *anywhere*.  What this approach solves is the problem to have it available in classes that inherit from arbitrary widgets or non-gui classes.

There are some preconfigured classes that can be used in various parts of an extension, but *any* class, be it a widget or any other class can additionally inherit from `ExtensionMixin` to "attach" itself to the extension (usually for direct access to the `settings()` and other properties.

Does that help?

Comment (PeterBjuhr) [reply frescobaldi_app/extensions/__init__.py]: Thanks! I have not a clear overview of this (yet). It just seemed to me to be an unusual design choice.

**Comment by uliska** [comment]
@wbsoft I think this PR should be considered beta quality code, but I suggest to merge it now. I think it will only be possible to iron out remaining flaws or determine remaining features that are necessary by creating some real-world extensions. And this will only be possible when the API is at least in the master branch.

What do you think? I would prefer having some feedback on this understanding before I merge the branch.

**Comment by uliska** [comment]
OK, after two days I finally merged this, knowing that there will probably be more work down the road...

**Pull request merged**